File: apps/ios/PhrenKit/Sources/PhrenKit/Search/SearchIndex.swift
import Foundation

/// On-device full-text search over the parsed snapshot. A simple in-memory
/// inverted index: at phren's documented scale (docs/performance.md treats
/// <1K findings as "small") a full rebuild per sync is milliseconds.
/// Archived content is excluded, matching the CLI's FTS index behavior.
public struct SearchIndex: Sendable {
    public enum DocKind: String, CaseIterable, Sendable {
        case finding, note, task, summary, truth
    }

    public struct Result: Identifiable, Equatable, Sendable {
        public let id: String
        /// Store the item came from ("" in single-store contexts).
        public let store: String
        public let project: String
        public let kind: DocKind
        public let text: String
        public let date: String?
        /// Finding type tag (for the type filter), when applicable.
        public let typeTag: String?
        public let score: Double
    }

    private struct Doc {
        let id: String
        let store: String
        let project: String
        let kind: DocKind
        let text: String
        let date: String?
        let recencyDate: Date?
        let typeTag: String?
        let tokens: [String: Int]
    }

    private var docs: [Doc] = []

    public init() {}

    public init(snapshot: LocalStore.Snapshot) {
        self.init(snapshots: [(store: "", snapshot: snapshot)])
    }

    public init(snapshots: [(store: String, snapshot: LocalStore.Snapshot)]) {
        // Dates are immutable for this index. Parse them once with a shared
        // formatter; only their age changes between searches.
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(identifier: "UTC")
        formatter.dateFormat = "yyyy-MM-dd"

        func doc(id: String, store: String, project: String, kind: DocKind,
                 text: String, date: String?, typeTag: String?) -> Doc {
            let recencyDate = date.flatMap { $0.count == 10 ? formatter.date(from: $0) : nil }
            return Doc(id: id, store: store, project: project, kind: kind, text: text, date: date,
                       recencyDate: recencyDate, typeTag: typeTag, tokens: Self.tokenFrequencies(text))
        }

        var docs: [Doc] = []
        for (store, snapshot) in snapshots {
            for (project, findings) in snapshot.findings {
                for finding in findings where !finding.archived {
                    docs.append(doc(
                        id: "f:\(store):\(project):\(finding.stableId ?? finding.id)",
                        store: store, project: project, kind: .finding, text: finding.text,
                        date: finding.date, typeTag: finding.typeTag
                    ))
                }
            }
            for (project, notes) in snapshot.notes {
                for note in notes {
                    docs.append(doc(
                        id: "n:\(store):\(project):\(note.stableId)",
                        store: store, project: project, kind: .note, text: note.text,
                        date: note.date, typeTag: nil
                    ))
                }
            }
            for (project, taskDoc) in snapshot.tasks {
                for task in taskDoc.allItems {
                    docs.append(doc(
                        id: "t:\(store):\(project):\(task.stableId ?? task.id)",
                        store: store, project: project, kind: .task, text: task.line,
                        date: task.createdAt.map { String($0.prefix(10)) }, typeTag: nil
                    ))
                }
            }
            // Truths are the *most* live knowledge in a store — pinned,
            // always injected by the CLI, never decayed — so they belong in
            // the index for exactly the reason archived findings don't.
            for (project, truths) in snapshot.truths {
                for truth in truths {
                    docs.append(doc(
                        id: "p:\(store):\(project):\(truth.id)",
                        store: store, project: project, kind: .truth, text: truth.text,
                        date: truth.addedDate, typeTag: nil
                    ))
                }
            }
            for (project, summary) in snapshot.summaries {
                for (i, paragraph) in summary.components(separatedBy: "\n\n").enumerated() {
                    let trimmed = paragraph.jsTrimmed
                    guard !trimmed.isEmpty, !trimmed.hasPrefix("#") else { continue }
                    docs.append(doc(
                        id: "s:\(store):\(project):\(i)",
                        store: store, project: project, kind: .summary, text: trimmed,
                        date: nil, typeTag: nil
                    ))
                }
            }
        }
        self.docs = docs
    }

    static func tokenize(_ text: String) -> [String] {
        text.lowercased()
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { $0.count >= 2 }
    }

    private static func tokenFrequencies(_ text: String) -> [String: Int] {
        var frequencies: [String: Int] = [:]
        for token in tokenize(text) {
            frequencies[token, default: 0] += 1
        }
        return frequencies
    }

    /// Term-frequency scoring with prefix matching on the final query token
    /// (so search feels instant while typing) and a recency boost by date.
    public func search(_ query: String, store: String? = nil, project: String? = nil,
                       kind: DocKind? = nil, typeTag: String? = nil,
                       limit: Int = 50) -> [Result] {
        search(query, store: store, project: project, kind: kind, typeTag: typeTag, limit: limit, now: Date())
    }

    /// One timestamp scores every match consistently and lets tests advance
    /// time without rebuilding the index or depending on the wall clock.
    func search(_ query: String, store: String? = nil, project: String? = nil,
                kind: DocKind? = nil, typeTag: String? = nil,
                limit: Int = 50, now: Date) -> [Result] {
        let queryTokens = Self.tokenize(query)
        guard !queryTokens.isEmpty else { return [] }

        var results: [Result] = []
        for doc in docs {
            if let store, doc.store != store { continue }
            if let project, doc.project != project { continue }
            if let kind, doc.kind != kind { continue }
            if let typeTag, doc.typeTag != typeTag { continue }

            var score = 0.0
            var matchedAll = true
            for (i, token) in queryTokens.enumerated() {
                let isLast = i == queryTokens.count - 1
                if let tf = doc.tokens[token] {
                    score += Double(tf)
                } else if isLast, doc.tokens.keys.contains(where: { $0.hasPrefix(token) }) {
                    score += 0.5
                } else {
                    matchedAll = false
                    break
                }
            }
            guard matchedAll, score > 0 else { continue }

            // Recency boost: newer date headings rank higher.
            if let date = doc.recencyDate {
                score += Self.recencyBoost(date, now: now)
            }
            results.append(Result(
                id: doc.id, store: doc.store, project: doc.project, kind: doc.kind,
                text: doc.text, date: doc.date, typeTag: doc.typeTag, score: score
            ))
        }
        return results.sorted { $0.score > $1.score }.prefix(limit).map { $0 }
    }

    private static func recencyBoost(_ date: Date, now: Date) -> Double {
        let ageDays = max(0, now.timeIntervalSince(date) / 86_400)
        return max(0, 2.0 - ageDays / 90.0)
    }
}

File: apps/android/phrenkit/src/main/kotlin/com/phren/kit/SearchIndex.kt
package com.phren.kit

import java.time.LocalDate
import java.time.ZoneOffset
import java.time.temporal.ChronoUnit
import kotlin.math.max

/**
 * On-device inverted index over live knowledge (port of SearchIndex.swift).
 * Archived (cold-tier) findings never enter it, matching the CLI.
 */
class SearchIndex private constructor(private val docs: List<Doc>) {
    enum class DocKind(val rawValue: String) { FINDING("finding"), NOTE("note"), TASK("task"), SUMMARY("summary"), TRUTH("truth") }

    data class Result(
        val id: String,
        val store: String,
        val project: String,
        val kind: DocKind,
        val text: String,
        val date: String?,
        val typeTag: String?,
        val score: Double,
    )

    private class Doc(
        val id: String,
        val store: String,
        val project: String,
        val kind: DocKind,
        val text: String,
        val date: String?,
        val typeTag: String?,
        val tokens: Map<String, Int>,
    )

    constructor() : this(emptyList())
    constructor(snapshot: LocalStore.Snapshot) : this(build(listOf("" to snapshot)))

    fun search(
        query: String,
        store: String? = null,
        project: String? = null,
        kind: DocKind? = null,
        typeTag: String? = null,
        limit: Int = 50,
    ): List<Result> {
        val queryTokens = tokenize(query)
        if (queryTokens.isEmpty()) return emptyList()
        val results = mutableListOf<Result>()
        for (doc in docs) {
            if (store != null && doc.store != store) continue
            if (project != null && doc.project != project) continue
            if (kind != null && doc.kind != kind) continue
            if (typeTag != null && doc.typeTag != typeTag) continue
            var score = 0.0
            var matchedAll = true
            for ((i, token) in queryTokens.withIndex()) {
                val isLast = i == queryTokens.size - 1
                val tf = doc.tokens[token]
                if (tf != null) score += tf
                else if (isLast && doc.tokens.keys.any { it.startsWith(token) }) score += 0.5
                else { matchedAll = false; break }
            }
            if (!matchedAll || score <= 0) continue
            // Recency boost: newer date headings rank higher.
            val date = doc.date
            if (date != null && date.length == 10) score += recencyBoost(date)
            results += Result(doc.id, doc.store, doc.project, doc.kind, doc.text, doc.date, doc.typeTag, score)
        }
        return results.sortedByDescending { it.score }.take(limit)
    }

    companion object {
        fun of(snapshots: List<Pair<String, LocalStore.Snapshot>>) = SearchIndex(build(snapshots))

        private fun build(snapshots: List<Pair<String, LocalStore.Snapshot>>): List<Doc> {
            val docs = mutableListOf<Doc>()
            fun add(id: String, store: String, project: String, kind: DocKind, text: String, date: String?, typeTag: String?) {
                docs += Doc(id, store, project, kind, text, date, typeTag, tokenFrequencies(text))
            }
            for ((store, snapshot) in snapshots) {
                for ((project, findings) in snapshot.findings) {
                    for (f in findings) if (!f.archived) {
                        add("f:$store:$project:${f.stableId ?: f.id}", store, project, DocKind.FINDING, f.text, f.date, f.typeTag)
                    }
                }
                for ((project, notes) in snapshot.notes) for (n in notes) {
                    add("n:$store:$project:${n.stableId}", store, project, DocKind.NOTE, n.text, n.date, null)
                }
                for ((project, taskDoc) in snapshot.tasks) for (t in taskDoc.allItems) {
                    add("t:$store:$project:${t.stableId ?: t.id}", store, project, DocKind.TASK, t.line, t.createdAt?.take(10), null)
                }
                // Truths are the most live knowledge in a store.
                for ((project, truths) in snapshot.truths) for (t in truths) {
                    add("p:$store:$project:${t.id}", store, project, DocKind.TRUTH, t.text, t.addedDate, null)
                }
                for ((project, summary) in snapshot.summaries) {
                    summary.split("\n\n").forEachIndexed { i, paragraph ->
                        val trimmed = paragraph.jsTrimmed
                        if (trimmed.isNotEmpty() && !trimmed.startsWith("#")) {
                            add("s:$store:$project:$i", store, project, DocKind.SUMMARY, trimmed, null, null)
                        }
                    }
                }
            }
            return docs
        }

        internal fun tokenize(text: String): List<String> {
            val out = mutableListOf<String>()
            val sb = StringBuilder()
            var i = 0
            val lower = text.lowercase()
            while (i < lower.length) {
                val cp = lower.codePointAt(i)
                if (Character.isLetterOrDigit(cp)) sb.appendCodePoint(cp)
                else {
                    if (sb.codePointCount(0, sb.length) >= 2) out += sb.toString()
                    sb.setLength(0)
                }
                i += Character.charCount(cp)
            }
            if (sb.codePointCount(0, sb.length) >= 2) out += sb.toString()
            return out
        }

        private fun tokenFrequencies(text: String): Map<String, Int> =
            tokenize(text).groupingBy { it }.eachCount()

        private fun recencyBoost(date: String): Double {
            val parsed = try { LocalDate.parse(date) } catch (_: Exception) { return 0.0 }
            val ageDays = max(0L, ChronoUnit.DAYS.between(parsed, LocalDate.now(ZoneOffset.UTC))).toDouble()
            return max(0.0, 2.0 - ageDays / 90.0)
        }
    }
}
